relay-server/src/actors/spooler/mod.rs

```diff
@@ -345,7 +349,19 @@ impl OnDisk {
             // to pick up.
             //
             // Right now we use 100 for batch size.
-            let mut envelopes = sql::delete_and_fetch(key, 100).fetch(&self.db).peekable();
+            let batch_size = 100;
+
+            // Before querying the db, make sure that the buffer guard has enough availability:
+            let mut attempts = 1;
+            while self.buffer_guard.used() > self.buffer_guard.capacity() / 2 {
+                tokio::time::sleep(Duration::from_millis(100)).await;
+                attempts += 1;
+            }
```

Comment:
I do not really like that we have the loop here running and burning the CPU, what if we use something like [`Notify`](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html) from within the `buffer_guard`, since it knows how many permits currently issued and dropped, we just wait for notification in this place, and continue once we are notified. 

something like this (pseudocode):
```rust
if self.buffer_guard.used() > self.buffer_guard.capacity() / 2 {
    self. buffer_guard.availability.notified().await;
}
```

And the metric will be not the number of attempts , but the time spent waiting on the notification.